### PR TITLE
app-emulation/xen: bump to 4.16.0

### PR DESCRIPTION
**Pull request by hydrapolic**
Compile tested on ~amd64.

```shell
--- xen-4.15.1-r2.ebuild        2021-12-19 11:41:47.711820349 +0000
+++ xen-4.16.0.ebuild   2021-12-20 20:18:22.015379812 +0000
@@ -1,9 +1,9 @@
 # Copyright 1999-2021 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=7
+EAPI=8
 
-PYTHON_COMPAT=( python3_{8..9} )
+PYTHON_COMPAT=( python3_{8..10} )
 
 inherit flag-o-matic mount-boot python-any-r1 toolchain-funcs
 
@@ -16,7 +16,7 @@
        SRC_URI=""
 else
        KEYWORDS="~amd64 ~arm -x86"
-       UPSTREAM_VER=2
+       UPSTREAM_VER=
        SECURITY_VER=
        GENTOO_VER=
 
@@ -95,7 +95,7 @@
        [[ -n ${GENTOO_VER} ]] && eapply "${WORKDIR}"/patches-gentoo
 
        # Symlinks do not work on fat32 volumes
-       eapply "${FILESDIR}"/${PN}-4.15-efi.patch
+       eapply "${FILESDIR}"/${PN}-4.16-efi.patch
 
        # Enable XSM-FLASK
        use flask && eapply "${FILESDIR}"/${PN}-4.15-flask.patch
```

```shell
--- xen-tools-4.15.1-r2.ebuild  2021-12-19 11:41:47.709820338 +0000
+++ xen-tools-4.16.0.ebuild     2021-12-20 20:18:22.015379812 +0000
@@ -3,7 +3,7 @@
 
 EAPI=7
 
-PYTHON_COMPAT=( python3_{8..9} )
+PYTHON_COMPAT=( python3_{8..10} )
 PYTHON_REQ_USE='ncurses,xml,threads(+)'
 
 inherit bash-completion-r1 flag-o-matic multilib python-single-r1 toolchain-funcs
@@ -17,21 +17,21 @@
        S="${WORKDIR}/${REPO}"
 else
        KEYWORDS="~amd64 ~arm ~arm64 ~x86"
-       UPSTREAM_VER=1
+       UPSTREAM_VER=
        SECURITY_VER=
        # xen-tools's gentoo patches tarball
-       GENTOO_VER=23
+       GENTOO_VER=24
        # xen-tools's gentoo patches version which apply to this specific ebuild
        GENTOO_GPV=0
        # xen-tools ovmf's patches
        OVMF_VER=
 
        SEABIOS_VER="1.14.0"
-       EDK2_COMMIT="a3741780fe3535e19e02efa869a7cac481891129"
+       EDK2_COMMIT="7b4a99be8a39c12d3a7fc4b8db9f0eab4ac688d5"
        EDK2_OPENSSL_VERSION="1_1_1j"
        EDK2_SOFTFLOAT_COMMIT="b64af41c3276f97f0e181920400ee056b9c88037"
        EDK2_BROTLI_COMMIT="666c3280cc11dc433c303d79a83d4ffbdd12cc8d"
-       IPXE_COMMIT="988d2c13cdf0f0b4140685af35ced70ac5b3283c"
+       IPXE_COMMIT="3c040ad387099483102708bb1839110bc788cefb"
 
        [[ -n ${UPSTREAM_VER} ]] && \
                UPSTREAM_PATCHSET_URI="https://dev.gentoo.org/~dlan/distfiles/${P/-tools/}-upstream-patches-${UPSTREAM_VER}.tar.xz
@@ -124,6 +124,8 @@
                system-seabios? ( sys-firmware/seabios )
                system-ipxe? ( sys-firmware/ipxe[qemu] )
                rombios? ( sys-devel/bin86 sys-devel/dev86 ) )
+       arm64? ( sys-power/iasl
+               rombios? ( sys-devel/bin86 sys-devel/dev86 ) )
        doc? (
                app-text/ghostscript-gpl
                app-text/pandoc
@@ -136,6 +138,7 @@
        hvm? ( x11-base/xorg-proto )
        qemu? (
                app-arch/snappy:=
+               dev-util/meson
                sdl? (
                        media-libs/libsdl[X]
                        media-libs/libsdl2[X]
@@ -154,12 +157,19 @@
 # Approved by QA team in bug #144032
 QA_WX_LOAD="
        usr/libexec/xen/boot/hvmloader
+       usr/libexec/xen/boot/ovmf.bin
        usr/libexec/xen/boot/xen-shim
        usr/share/qemu-xen/qemu/hppa-firmware.img
+       usr/share/qemu-xen/qemu/opensbi-riscv32-generic-fw_dynamic.elf
+       usr/share/qemu-xen/qemu/opensbi-riscv64-generic-fw_dynamic.elf
        usr/share/qemu-xen/qemu/s390-ccw.img
        usr/share/qemu-xen/qemu/u-boot.e500
 "
 
+QA_EXECSTACK="
+       usr/share/qemu-xen/qemu/hppa-firmware.img
+"
+
 QA_PREBUILT="
        usr/libexec/xen/bin/elf2dmp
        usr/libexec/xen/bin/ivshmem-client
@@ -173,6 +183,7 @@
        usr/libexec/xen/bin/qemu-storage-daemon
        usr/libexec/xen/bin/qemu-system-i386
        usr/libexec/xen/bin/virtfs-proxy-helper
+       usr/libexec/xen/boot/ovmf.bin
        usr/libexec/xen/boot/xen-shim
        usr/libexec/xen/libexec/qemu-pr-helper
        usr/libexec/xen/libexec/virtfs-proxy-helper
@@ -185,8 +196,6 @@
 
 RESTRICT="test"
 
-PATCHES=( "${FILESDIR}/${PN}-4.15.0-fix-xenstat-python-bindings.patch" )
-
 pkg_setup() {
        python_setup
        export "CONFIG_LOMOUNT=y"
@@ -293,18 +302,11 @@
                pushd tools/firmware/ovmf-dir-remote/MdeModulePkg/Library/BrotliCustomDecompressLib/brotli > /dev/null
                        eapply "${FILESDIR}/${PN}-4.15.1-brotli-gcc11.patch"
                popd > /dev/null
-
-               pushd tools/firmware/ovmf-dir-remote > /dev/null
-                       eapply "${FILESDIR}/${PN}-4.15.1-edk2-python3.9.patch"
-               popd > /dev/null
        fi
 
-       # Fix building with ocaml 4.12 #818100
-       eapply "${FILESDIR}/${PN}-4.15.1-ocaml-4.12.patch"
-
        # ipxe
        if use ipxe; then
-               cp "${DISTDIR}/ipxe-git-${IPXE_COMMIT}.tar.gz" tools/firmware/etherboot/_ipxe.tar.gz || die
+               cp "${DISTDIR}/ipxe-git-${IPXE_COMMIT}.tar.gz" tools/firmware/etherboot/ipxe.tar.gz || die
 
                # gcc 11
                cp "${WORKDIR}/patches-gentoo/${PN}-4.15.0-ipxe-gcc11.patch" tools/firmware/etherboot/patches/ipxe-gcc11.patch || die
```

### CODE REVIEW

File: dev-python/pypam/pypam-0.5.0-r7.ebuild [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Copyright 1999-2021 Gentoo Authors
```
Comment (thesamesam) [inline dev-python/pypam/pypam-0.5.0-r7.ebuild:1]: Did you mean to include this in this PR? :)

Comment (thesamesam) [reply dev-python/pypam/pypam-0.5.0-r7.ebuild:1]: Oh, I see you did! By the way, I did notice you fixed the || die before, I just forgot/lost the open tab

Comment (hydrapolic) [reply dev-python/pypam/pypam-0.5.0-r7.ebuild:1]: This was needed to add python 3.10 to xen-tools.

**Comment by hydrapolic** [comment]
Thank you @thesamesam

**Pull request closed without merging**